src/components/constraints/ConstraintForm.svelte

```diff
@@ -3,233 +3,376 @@
-              : featurePermissions.constraints.canUpdate(user, plan))
-          );
-        }
-        return prevPermission;
-      }, false);
-    }
-
-    return true;
-  }
-
-  function hasPlanPermission(plan: PlanSlim, mode: 'create' | 'edit', user: User | null): boolean {
-    if (user) {
-      return mode === 'create'
-        ? featurePermissions.constraints.canCreate(user, plan)
-        : featurePermissions.constraints.canUpdate(user, plan);
+  $: if (constraintPublic && constraintName) {
+    const existingMetadata = $constraints.find(
+      ({ name, public: publicConstraint }) => name === constraintName && publicConstraint,
+    );
+    if (existingMetadata !== undefined && existingMetadata.id !== constraintMetadataId) {
+      constraintNameError = 'Constraint name must be unique when public';
+    } else {
+      constraintNameError = '';
     }
-    return false;
+  } else {
+    constraintNameError = '';
   }
 
-  function diffConstraints(constraintA: Partial<Constraint>, constraintB: Partial<Constraint>) {
+  function diffConstraintMetadata(
+    constraintMetadataA: SavedConstraintMetadata,
+    constraintMetadataB: SavedConstraintMetadata,
+  ) {
     if (
-      constraintA.definition !== constraintB.definition ||
-      constraintA.description !== constraintB.description ||
-      constraintA.name !== constraintB.name ||
-      constraintA.plan_id !== constraintB.plan_id ||
+      constraintMetadataA.description !== constraintMetadataB.description ||
+      constraintMetadataA.name !== constraintMetadataB.name ||
+      constraintMetadataA.public !== constraintMetadataB.public ||
+      constraintMetadataA.owner !== constraintMetadataB.owner ||
       diffTags(
-        (constraintA.tags || []).map(({ tag }) => tag),
-        (constraintB.tags || []).map(({ tag }) => tag),
+        (constraintMetadataA.tags || []).map(({ tag }) => tag),
+        (constraintMetadataB.tags || []).map(({ tag }) => tag),
       )
     ) {
       return true;
-    } else if (constraintA.plan_id === null && constraintB.plan_id === null) {
-      // only diff model_id if both plan_ids are null
-      // to replicate the behavior where when saving a constraint, the model_id is ignored
-      // if a plan_id is supplied
-      return constraintA.model_id !== constraintB.model_id;
     }
+
+    return false;
+  }
+
+  function diffConstraintDefinition(
+    constraintDefinitionA: Pick<SavedConstraintDefinition, 'definition'>,
+    constraintDefinitionB: Pick<SavedConstraintDefinition, 'definition'>,
+  ) {
+    return constraintDefinitionA.definition !== constraintDefinitionB.definition;
   }
 
   function onDidChangeModelContent(event: CustomEvent<{ value: string }>) {
     const { detail } = event;
     const { value } = detail;
-    constraintDefinition = value;
+    constraintDefinitionCode = value;
   }
 
-  function onKeydown(event: KeyboardEvent): void {
-    if (isSaveEvent(event)) {
```

Comment:
No more control/command + s?

Reply:
I kind of wanted to discourage being able to save too easily since it'll be potentially creating a new version every time. I'm open to adding it back though.

Reply:
Good thinking, agreed.